tests/integration/rolling_update/test_rolling_update.py

```diff
@@ -16,52 +19,82 @@ def config(tmpdir):
     del os.environ['JINA_REPLICA_DIR']
 
 
-def get_doc(i):
-    return Document(text=f'doc {i}', embedding=np.array([i] * 5))
+@pytest.fixture
+def docs():
+    return [
+        Document(id=str(i), text=f'doc {i}', embedding=np.array([i] * 5))
+        for i in range(20)
+    ]
+
+
+class DummyMarkExecutor(BaseEncoder):
+    def get_docs(self, req_type):
+        if req_type == 'ControlRequest':
+            return []
+        driver = self._drivers[req_type][0]
+        return driver.docs
+
+    def __call__(self, req_type, *args, **kwargs):
+        if req_type == 'ControlRequest':
+            for d in self._drivers[req_type]:
+                d()
+        else:
+            for doc in self.get_docs(req_type):
+                doc.tags['replica'] = self.replica_id
+                doc.tags['shard'] = self.pea_id
 
 
-def test_normal(config):
+def test_normal(docs):
+    NUM_REPLICAS = 3
+    NUM_SHARDS = 2
     # this test is a bit hacky.
     # It uses the score field to pass the information of the used replica during search.
     # Please don't use it that way in application code
-    used_replicas = []
+    doc_id_path = collections.OrderedDict()
 
     def handle_search_result(resp):
-        used_replicas.append(list(resp.search.docs)[0].matches[0].score.value)
+        for doc in resp.search.docs:
+            doc_id_path[int(doc.id)] = (doc.tags['replica'], doc.tags['shard'])
 
     flow = Flow().add(
         name='pod1',
-        uses=os.path.join(cur_dir, 'yaml/mock_index_vector.yml'),
-        replicas=3,
-        parallel=2,
+        uses='!DummyMarkExecutor',
+        replicas=NUM_REPLICAS,
+        parallel=NUM_SHARDS,
     )
     with flow:
-        for i in range(20):
-            # test rolling update does not hang
-            flow.search(get_doc(0), on_done=handle_search_result)
+        flow.search(inputs=docs, request_size=1, on_done=handle_search_result)
+
+    assert len(doc_id_path.keys()) == len(docs)
+
+    num_used_replicas = len(set(map(lambda x: x[0], doc_id_path.values())))
+    assert num_used_replicas == NUM_REPLICAS
+
+    shards = collections.defaultdict(list)
+    for replica, shard in doc_id_path.values():
+        shards[replica].append(shard)
 
-    # 20 time one of the replicas is called
-    assert len(used_replicas) == 20
+    assert len(shards.keys()) == NUM_REPLICAS
 
-    # there are three replicas in total
-    assert set(used_replicas) == {0.0, 1.0, 2.0}
+    for shard_list in shards.values():
+        assert len(set(shard_list)) == NUM_SHARDS
 
 
 @pytest.mark.timeout(30)
-def test_simple_run():
+def test_simple_run(docs):
     flow = Flow().add(
         name='pod1',
         replicas=2,
         parallel=3,
     )
     with flow:
         # test rolling update does not hang
-        flow.search(get_doc(0))
+        flow.search(docs)
         flow.rolling_update('pod1')
-        flow.search(get_doc(1))
+        flow.search(docs)
```

Comment:
should we validate results?

Reply:
for now it just tests it is not blocking, but there is a bug there. Will work on that in another iteration.

Related to #2255